### ISSUE TITLE
History plot on the compare benchmark result page?

### ISSUE THREAD

**Issue by jonkeane** [conbench/conbench#1209]
We link to the [compare/runs page](https://conbench.ursa.dev/compare/runs/536d3a2929894f629560e66253bb9418...8d0be02fd53c451985848e1288179e55/) when there are regressions from benchalerts (+ also the bespoke Arrow alerts). This then links to [compare/benchmarks (results) page](https://conbench.ursa.dev/compare/benchmarks/9e32fc4c91274dac8012cb5531e069b6...45789e2ee72a402caa1250144272195c/) for a single benchmark result. But from that page, it's not totally obvious how I could see the historical data so that I could see if this is really a true regression, something that will end up just being noise, or what.

One possible solution here is to add the history plot to this page.

A workaround that exists (but is not at all obvious): one can click the contender link up top and that takes you to the contender page which has the history plot. But in order to know to do that you have to know quite a bit about conbench generally. And that's yet another click to get to the data one is interested in.

**Comment by jgehrcke** [comment]
Yes! Thank you, Jon.

I also forgot to create a ticket earlier based on user feedback on Slack ( :-) @thisisnic)

Some quotes from Nic and myself on the thread

> I expected to see an interactive plot of commits at the top, but it's not there

> In those direct comparison views, we find links to the history for the individual benchmark scenarios. Those are most insightful I think.

> Now I see the path from the notifications to where I want to be (i.e. the history plot), I think I have everything I need here! 

> We will need to make this "path from the notifications to where I want to be" better for sure.

**Comment by jgehrcke** [comment]
Thought about this a little more.

> that takes you to the contender page which has the history plot. But in order to know to do that you have to know quite a bit about conbench generally. And that's yet another click to get to the data one is interested in.

And even if you do so, you do not get to see the two specific benchmark results that you actually wanted to compare in wider context: baseline and contender. All you get is a history plot showing a "current benchmark result", and a smart user might remember that "current benchmark result" is contender or baseline, and they might go actively look for the other one in the history plot by clicking through data points. Really bad UX.

We also cannot just "add the history plot to this page". To be considered: The "compare benchmark result page" allows for comparing a wider range of benchmark results than the history plot is (rightfully) designed to show results for. 

- The history plot requires a set of properties H to be equivalent for two benchmark results to be "history-comparable", i.e. to be shown in the same plot.
- The benchmark result comparison page has a set of properties C that it requires to be equivalent before it shows a comparison.

H is vastly different from C, and so there needs to be more logic and user feedback. I just tested this out, the compare page even allows for comparing benchmark results from different benchmarks:

![image](https://user-images.githubusercontent.com/265630/236616803-79889bb6-3c5a-4752-8986-2cdcffb83807.png)

Here's what I think what we should do:
1) In the benchmark result view (which shows the history plot), allow for putting in an argument: another benchmark result ID. If provided and if part of history, highlight it in plot (in addition to the current benchmark result).
2) In code, make a nice abstraction for BenchmarkResult like `bmr1.is_history_comparable(bmr2)`. Boolean outcome. Can use that in the benchmark result compare view. If this is True, then put a prominent section into the page: "these two results are history-comparable, go _here_ to see both results in their wider history context".

**Comment by jgehrcke** [comment]
Based on the proposal above, I started looking into this. And then went right into a code section that I worked in before ... and remembered :) The right thing to do here would be to say "we cannot do this until we remove the API layer indirection from the /compare UI endpoints", i.e. do parts of https://github.com/conbench/conbench/issues/968. 

Mapping this 'resolve this particularly serious tech debt item' out as a hard dependency, no feelings attached, normal procedure.

To give a bit of an impression, imagine making changes here:

https://github.com/conbench/conbench/blob/95d40c938500e1fdd169aafb06a0ee7754cb8a74/conbench/app/compare.py#L32

or here:
https://github.com/conbench/conbench/blob/95d40c938500e1fdd169aafb06a0ee7754cb8a74/conbench/app/compare.py#L213

The code comments are already mine, from having worked through the hardest parts. But even with these comments it's still too hard. 

Last time I already put in a brutal hack in there where we are fetching a benchmark result a second time from the DB (after having gone through the API layer indirection, basically completely ignoring this indirection, but also not removing it): https://github.com/conbench/conbench/blob/95d40c938500e1fdd169aafb06a0ee7754cb8a74/conbench/app/compare.py#L60 just because it was easier to get hands on data we can reason about.
